envision/data_formatter.py

```diff
@@ -0,0 +1,364 @@
+    ):
+        # self.seen_objects = context.seen_objects if context else set()
+        self.id: Any = id
+        self._data: List[Any] = []
+        self._reduction_context = ReductionContext(
+            enabled=formatter_params.enable_reduction
+        )
+        self._serializer = formatter_params.serializer
+        self._float_decimals = formatter_params.float_decimals
+        self._bool_as_int = formatter_params.bool_as_int
+
+    def reset(self, reset_reduction_context: bool = True):
+        """Reset the current context in preparation for new serialization."""
+        self._data = []
+        if reset_reduction_context:
+            self._reduction_context.reset()
+
+    def add_any(self, obj: Any):
+        """Format the given object to the current layer."""
+        if type(obj) in _primitives:
+            self.add_primitive(obj)
+        else:
+            self.add(obj, None)
+
+    def add_primitive(self, obj: Any):
+        """Add the given object as is to the given layer. Will decompose known primitives."""
+        # TODO prevent cycles
+        # if isinstance(ref, (Object)) and ref in self.seen_objects:
+        #     return
+
+        f = _formatter_map.get(type(obj))
+        if f:
+            f(obj, self)
+        else:
+            if isinstance(obj, float):
+                obj = round(obj, self._float_decimals)
+            elif self._bool_as_int and isinstance(obj, (bool, np.bool_)):
+                obj = int(obj)
+            self._data.append(obj)
+
+    def add(
+        self,
+        value: Any,
+        id_: Optional[str],
+        op: Operation = Operation.NONE,
+        alternate: Callable[[Any], Any] = lambda v: v,
+    ):
+        """Format the given object to the current layer. Specified operations are performed."""
+        outval = value
+        f = _sequence_formatter_map.get(type(value))
+        if op & Operation.REDUCE:
+            outval = self._reduction_context.resolve_value(outval)
+        if op & Operation.FLATTEN:
+            outval = unpack(outval)
+            if not isinstance(outval, (Sequence, np.ndarray)):
+                assert False, "Must use flatten with Sequence or dataclass"
+            for e in outval:
+                self.add_primitive(e)
+        elif f:
+            f(value, self)
+        else:
+            self.add_primitive(outval)
+
+    class DataFormatterLayer(ContextManager, Iterator):
+        """A formatting layer that maps into the above layer of the current data formatter."""
+
+        def __init__(
+            self,
+            data_formatter: "EnvisionDataFormatter",
+            iterable: Optional[Iterable],
+            op: Operation,
+        ) -> None:
+            super().__init__()
+            self._data_formatter = data_formatter
+            self._upper_layer_data = data_formatter._data
+            self._operation = op
+
+            def empty_gen():
+                return
+                yield
```

Comment:
The `yield` line is unreachable.

Reply:
This is intentional. It turns the method into a generator.